runner/logs.go

```diff
@@ -111,73 +313,126 @@ func TestLogsForAllServers(servers []*state.Server, globalCollectionOpts state.C
-func DownloadLogsFromAllServers(servers []*state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) {
-	var wg sync.WaitGroup
+func testLocalLogTail(ctx context.Context, wg *sync.WaitGroup, server *state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) bool {
+	logger.PrintInfo("Testing log collection (local)...")
 
-	if !globalCollectionOpts.CollectLogs {
-		return
+	logTestSucceeded := make(chan bool, 1)
+	parsedLogStream := setupLogStreamer(ctx, wg, globalCollectionOpts, logger, []*state.Server{server}, logTestSucceeded, stream.LogTestCollectorIdentify)
+
+	err := selfhosted.SetupLogTailForServer(ctx, wg, globalCollectionOpts, logger, server, parsedLogStream)
+	if err != nil {
+		logger.PrintError("ERROR - Could not tail logs for server: %s", err)
+		return false
 	}
 
-	for _, server := range servers {
-		if server.Config.DisableLogs || (server.Grant.Valid && !server.Grant.Config.EnableLogs) {
-			continue
-		}
+	logs.EmitTestLogMsg(server, globalCollectionOpts, logger)
 
-		if server.Config.AwsDbInstanceID == "" {
-			continue
-		}
+	select {
+	case <-logTestSucceeded:
+		break // TODO: Is this correct?
+	case <-time.After(10 * time.Second):
+		logger.PrintError("ERROR - Local log tail timed out after 10 seconds - did not find expected log event in stream")
+		return false
+	}
+
+	logger.PrintInfo("  Local log test successful")
+	return true
+}
+
+func testRdsLogDownload(ctx context.Context, wg *sync.WaitGroup, server *state.Server, globalCollectionOpts state.CollectionOpts, prefixedLogger *util.Logger) bool {
+	prefixedLogger.PrintInfo("Testing log collection (Amazon RDS)...")
+	_, _, err := downloadLogsForServer(server, globalCollectionOpts, prefixedLogger)
+	if err != nil {
+		prefixedLogger.PrintError("ERROR - Could not get Amazon RDS logs: %s", err)
+		return false
+	}
+
+	prefixedLogger.PrintInfo("  Log test successful")
+	return true
+}
+
+func testAzureLogStream(ctx context.Context, wg *sync.WaitGroup, server *state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) bool {
+	logger.PrintInfo("Testing log collection (Azure Database)...")
+
+	logTestSucceeded := make(chan bool, 1)
+	parsedLogStream := setupLogStreamer(ctx, wg, globalCollectionOpts, logger, []*state.Server{server}, logTestSucceeded, stream.LogTestAnyEvent)
 
-		wg.Add(1)
-		go downloadLogsFromOneServer(&wg, server, globalCollectionOpts, logger)
+	err := azure.SetupLogSubscriber(ctx, wg, globalCollectionOpts, logger, []*state.Server{server}, parsedLogStream)
+	if err != nil {
+		logger.PrintError("ERROR - Could not get logs through Azure Event Hub: %s", err)
+		return false
 	}
 
-	wg.Wait()
+	logs.EmitTestLogMsg(server, globalCollectionOpts, logger)
 
-	return
+	select {
+	case <-logTestSucceeded:
+		break // TODO: Is this correct?
+	case <-time.After(10 * time.Second):
+		logger.PrintError("ERROR - Azure Event Hub log tail timed out after 10 seconds - did not receive any log events")
```

Comment:
I guess with Azure and GCP log tests, we still have the issue of potentially competing with a running collector on the subscription, right? Should we indicate that in the message here?

Reply:
Ah, thanks for the reminder, I forgot that failure scenario.

I'll add a hint message like the following: "HINT - This error may be a false positive if the collector is running in the background and consumes the same [provider name] stream" -- but we probably want to find a better way to do this in general.

*thinks about this briefly*

We could signal the collector running in the background (like we do for a reload) to not acknowledge the GCP Pub Sub messages - that way they should still be available for the test collector to be retrieved. e.g. when the collector receives a `SIGUSR1` signal it doesn't acknowledge any messages for the next 10 seconds. We'd have to test whether the acknowledge handling actually behaves that way, but that could work (at least for GCP where there is an explicit ack from the client).

Reply:
Ah interesting—I was thinking about signaling, too, but didn't think of using that to not ack. That could work, although maybe that would lead to duplicate messages? Maybe instead we could use the signal to somehow share the received messages (e.g., an ephemeral localhost web server?). But yeah, for now, I think just a HINT in the message works.